### ISSUE TITLE
matching Etag of file uploaded

### ISSUE THREAD

**Issue by nasirmehmood** [aws/aws-sdk-ios#208]
Hi,
I am creating md5 of the file being uploaded using method 
+ (NSString *)aws_base64md5FromData:(NSData *)data
from sdk 2.2.0

But the Etag I receive in the headers after successful upload is not matching with the above computed md5. Am I missing something here ?

Please provide response on it.
Thanks

**Comment by nasirmehmood** [comment]
Can anyone provide me guidance here please ? What do I need to do here to match the Etag with locally generated md5 ?

**Comment by yosuke-matsuda** [comment]
You are not supposed to use Etag for ensuring data integrity. You should set `contentMD5` when uploading data to Amazon S3.

**Comment by nasirmehmood** [comment]
Thanks for the response.
If I set 'Content-MD5' header then NSURLSessionUploadTask don't seem to work. It sends initial bytes and then gets stuck, sending initial bytes again and again. But if I don't set this header then uploading works fine.
I am assuming here, if 'Content-MD5' is set and request body contains the data (e.g. POST request) then Amazon server sends correct response. But for uploading task case, the request type is PUT, so in that case server is sending incorrect response which causes the process to stuck at initial bytes of uploading.

**Comment by yosuke-matsuda** [comment]
What do you mean by "`NSURLSessionUploadTask` don't seem to work?" Please share a code snippet to reproduce the issue.

**Comment by nasirmehmood** [comment]
I am uploading a large file in chunks of different sizes. So each chunk is written to a file, file is uploaded on server and when all done I combine them on server. 
Below is the code I am using for upload
```
- (void) uploadChunk:(NSMutableDictionary*)chunkInfo
{
    int chunkNumber = [[chunkInfo objectForKey:kChunkNumberKey] intValue];
    NSUInteger chunkSize = [[chunkInfo objectForKey:kChunkSizeKey] unsignedIntegerValue];
    unsigned long long position=[[chunkInfo objectForKey:kChunkOffsetKey] unsignedLongLongValue];
    
    [_fileHandle seekToFileOffset:position];
    
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    NSString *chunkFilePath=[documentsDirectory stringByAppendingPathComponent:[NSString stringWithFormat:@"chunk%@.bin", [chunkInfo objectForKey:kChunkNumberKey]]];
    
    NSData *chunkData=[_fileHandle readDataOfLength:chunkSize];
    [chunkData writeToFile:chunkFilePath atomically:NO];
    
    NSString *md5Base64 = [self aws_base64md5FromData:chunkData];
    chunkData=nil;
    
    NSString *fileContentTypeStr = @"application/octet-stream";
    NSURL *presignedURL = [NSURL URLWithString:[chunkInfo objectForKey:@"url"]];
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:presignedURL];
    request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
    [request setHTTPMethod:@"PUT"];
    [request setValue:fileContentTypeStr forHTTPHeaderField:@"Content-Type"];
    
    [request setValue:md5Base64 forHTTPHeaderField:@"Content-MD5"];
    [request setValue:[_requestOptions objectForKey:@"videoID"] forHTTPHeaderField:kVideoIDKey];
    [request setValue:[[_requestOptions objectForKey:@"videoPath"] lastPathComponent] forHTTPHeaderField:kFileNameKey];
    [request setValue:kMethodNameUploadChunk forHTTPHeaderField:kMethodNameKey];
    [request setValue:[NSString stringWithFormat:@"%d", chunkNumber] forHTTPHeaderField:kChunkNumberKey];
    [request setValue:chunkFilePath.lastPathComponent forHTTPHeaderField:kChunkFileNameKey];

	NSURLSessionUploadTask *uploadTask=[[self backgroundSession] uploadTaskWithRequest:request fromFile:[NSURL fileURLWithPath:chunkFilePath]];
    
    [uploadTask resume];
    [[self foregroundSession] finishTasksAndInvalidate];
}
```
And for md5 generation, I am using function from SDK 2.2.0
```
- (NSString *)aws_base64md5FromData:(NSData *)data {
    
    if([data length] > UINT32_MAX)
    {
        //The NSData size is too large. The maximum allowable size is UINT32_MAX.
        return nil;
    }
    
    const void    *cStr = [data bytes];
    unsigned char result[CC_MD5_DIGEST_LENGTH];
    
    CC_MD5(cStr, (uint32_t)[data length], result);
    
    NSData *md5 = [[NSData alloc] initWithBytes:result length:CC_MD5_DIGEST_LENGTH];
    return [md5 base64EncodedStringWithOptions:kNilOptions];
}
```

Uploading works fine if I don't set ``` Content-MD5 ``` here, but if I set, then the uploadingTask stucks at initial bytes e.g. here it uploads 0.16MB then stops and again starts uploading from 0MB. It continues this way, not proceeding any further.

It would be great if you can provide any help.
Thanks

**Comment by yosuke-matsuda** [comment]
It seems like you are not using the SDK to upload the data (just to calculate MD5). If you want to implement your own solution, [Amazon S3 Developer Guide](http://docs.aws.amazon.com/AmazonS3/latest/dev/Welcome.html) and [Amazon S3 API Reference](http://docs.aws.amazon.com/AmazonS3/latest/API/Welcome.html) are great resources to help you do that.

**Comment by nasirmehmood** [comment]
I have managed to find solution for it.
there is a method to get PreSignedURL is AWS SDK
```- (AWSTask *)getPreSignedURL:(AWSS3GetPreSignedURLRequest *)getPreSignedURLRequest```
here it takes a param ```AWSS3GetPreSignedURLRequest```
Inside this method, when creating ```signature```,  ```contentMd5``` is set and is always empty string. We can set  ```contentType``` but we can't set ```contentMd5``` to our value as ```AWSS3GetPreSignedURLRequest``` don't have any such property.
I added an additional param to this method to pass MD5 of my file, which now becomes part of the ```signature```. After that when I set ```Content-MD5``` header of the upload task request, it uploads fine and I get error response if the uploaded file is invalid.

I assume this method should be updated or ```contentMd5``` property should be added to ```AWSS3GetPreSignedURLRequest``` so we can pass our own MD5 to become part of the ```signature```.

Thanks for all your responses.
